Calculator/Calculator.cpp

```diff
@@ -82,97 +84,147 @@ Token Token_stream::get()
 
 //------------------------------------------------------------------------------
 
-Token_stream ts;       // provides get() and putback()
+Token_stream ts;       
 
 //------------------------------------------------------------------------------
 
-double expression();  // read and evaluate a Expression
+double expression();  
 
 //------------------------------------------------------------------------------
+// Gets next token from the token stream and evaluates expression grouped within
+// '{}' or '()' as well as single number expressions and returns the value 
+// for the grouped expression or single number
 
-double primary()     // read and evaluate a Primary
+double primary()     
 {
+    
     Token t = ts.get();
     switch (t.kind) {
-    case '(':    // handle '(' expression ')'
+    case '{':
+    {
+        double c = expression();
+        t = ts.get();
+        if (t.kind != '}') error("'}' expected");
+        return c;
+    }
+    case '(':    
     {
         double d = expression();
         t = ts.get();
         if (t.kind != ')') error("')' expected");
         return d;
     }
-    case 'n':            // we use 'n' to represent a number
-        return t.value;  // return the number's value
+    case 'n':         
+    {
+        return t.value;  
+    }
     default:
         error("primary expected");
     }
 }
 
 //------------------------------------------------------------------------------
+// Gets next token from token stream and calculates a factorial denoted by '!' and
+// returns the value for the factorial. All tokens other than '!' are putback into the
+// token stream
+
+double factorial() {
+    double left = primary();
+    Token t = ts.get();
+    while (true) {
+
+        switch (t.kind) {
+        case '!':
+        {
+            int x = left;
+            int k = (x - 1);
+            double f = x;
+            for (int i = 0; i < k; ++i) {
+                f *= (--x);
+            }
+            return f;
+        }
+        default:
+            ts.putback(t);
+            return left;
+        }
+    }
+}
+
```

Comment:
impressive really. even though the guts of how you calculate your factorial is truly fucked up...where you have it placed in the design as well as how it is handled is exactly like mine. no need to make any changes now. we can go over my final answer and you can get some ideas for future reference.